### PR TITLE
heap limits

### PR DESCRIPTION
**Pull request by arvidn**
and implement a heap- and atom size limits

### CODE REVIEW

File: src/limits.rs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+pub const MAX_CUMULATIVE_HEAP_SIZE: usize = 200 * 1024 * 1024;
+pub const MAX_ATOM_SIZE: usize = 1024 * 1024;
```
Comment (arvidn) [inline src/limits.rs:2]: I think these limits should be significantly higher than we would expect any normal program to need, does this seem like reasonable limits?

Comment (richardkiss) [reply src/limits.rs:2]: Don't use constants, make them parameters.

File: src/py/arc_allocator.rs [hunk 2]
```diff
@@ -38,10 +42,12 @@ impl Clone for ArcSExp {
 
 impl ArcAllocator {
     pub fn new() -> Self {
-        ArcAllocator {}
+        ArcAllocator {
+            cumulative_allocations: 0,
```
Comment (richardkiss) [inline src/py/arc_allocator.rs:46]: This won't work because `ArcAllocator` objects interoperate.

**Comment by richardkiss** [comment]
Hard limits not associated with costs are consensus risks and make it much more likely that different implementations will have hard-to-find differences, so I'm not a big fan of some of these changes.

File: src/py/arc_allocator.rs (hunk 2 above)
Comment (arvidn) [reply src/py/arc_allocator.rs:46]: there's more than one allocator?

**Comment by arvidn** [comment]
> Hard limits not associated with costs are consensus risks and make it much more likely that different implementations will have hard-to-find differences, so I'm not a big fan of some of these changes.

Right. However, the alternative (as far as I can tell) is to have costs of operations be proportional to the cumulative allocated memory so far. Otherwise the cost doesn't grow fast enough once you start using a lot of memory.

**Comment by arvidn** [comment]
Actually, the upper limit of atoms is trivial to implement and unlikely (I would say) to cause consensus divergence. The cumulative allocation is a bit easier to get wrong I suppose. As it turns out the `ArcAllocator` apparently can't keep state.

**Comment by richardkiss** [comment]
Yeah, allocators that use GC like `ArcAllocator` are going to have a very hard time with state across a given program. I think artificially increasing the cost of allocation is our best bet even though it might seem a bit wasteful. We should pick a maximum amount of memory we want to allocate and adjust our per-byte costs accordingly. Mostly `concat` – not sure it's possible to waste much memory with any other operators. Maybe arithmetic.

**Comment by arvidn** [comment]
I broke out, what I believe to be, the uncontroversial changes: https://github.com/Chia-Network/clvm_rs/pull/40
Thats just to support the error path from allocators.

**Pull request closed without merging**